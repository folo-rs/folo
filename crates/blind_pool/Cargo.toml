[package]
name = "blind_pool"
description = "A pool of pinned objects with opaque types - it can contain anything while pooling the memory for it"
publish = true
version = "0.1.0"

authors.workspace = true
edition.workspace = true
license.workspace = true
repository.workspace = true
rust-version.workspace = true

[features]
default = []

[dependencies]
dataless_pool = { workspace = true }
foldhash = { workspace = true }

[dev-dependencies]
criterion = { workspace = true }
mutants = { workspace = true }

[lints]
workspace = true
