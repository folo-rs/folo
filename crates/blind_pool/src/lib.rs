//! This crate provides [`BlindPool`], a dynamically growing pool of objects that can store
//! objects of any type.
//!
//! It offers stable memory addresses and efficient typed insertion with automatic value dropping.
//!
//! # Type-agnostic memory management
//!
//! [`BlindPool`] can accept any type and will automatically manage the appropriate
//! storage as needed.
//!
//! The pool itself does not hold or create any `&` shared or `&mut` exclusive references to its
//! contents, allowing the caller to decide who and when can obtain a reference to the inserted
//! values. The caller is responsible for ensuring that Rust aliasing rules are respected.
//!
//! # Features
//!
//! - **Type-agnostic memory management**: Accepts any type.
//! - **Stable addresses**: Memory addresses remain valid until explicitly removed.
//! - **Automatic dropping**: Values are properly dropped when removed from the pool.
//! - **Dynamic growth**: Pool capacity grows automatically as needed.
//! - **Efficient allocation**: Uses high density slabs to minimize allocation overhead.
//! - **Stable Rust**: No unstable Rust features required.
//! - **Optional leak detection**: Pool can be configured to panic on drop if values are still present.
//!
//! # Example
//!
//! ```rust
//! use blind_pool::BlindPool;
//!
//! // Create a blind pool that can store any type.
//! let mut pool = BlindPool::new();
//!
//! // Insert values of different types into the same pool.
//! let pooled_u64 = pool.insert(42_u64);
//! let pooled_i32 = pool.insert(-123_i32);
//! let pooled_f32 = pool.insert(3.14_f32);
//!
//! // Read data back from the pooled items.
//! let value_u64 = unsafe {
//!     // SAFETY: The pointer is valid and the value was just inserted.
//!     pooled_u64.ptr().read()
//! };
//!
//! let value_i32 = unsafe {
//!     // SAFETY: The pointer is valid and the value was just inserted.
//!     pooled_i32.ptr().read()
//! };
//!
//! assert_eq!(value_u64, 42);
//! assert_eq!(value_i32, -123);
//! ```

mod builder;
mod pool;

pub use builder::*;
// Re-export DropPolicy from opaque_pool
pub use opaque_pool::DropPolicy;
pub use pool::*;
