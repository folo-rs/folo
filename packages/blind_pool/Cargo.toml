[package]
name = "blind_pool"
description = "A pinned object pool that can store objects of any type"
publish = true
version = "0.4.6"

authors.workspace = true
edition.workspace = true
license.workspace = true
repository.workspace = true
rust-version.workspace = true

[features]
default = []

[dependencies]
foldhash = { workspace = true }
opaque_pool = { workspace = true }
pastey = { workspace = true }

[dev-dependencies]
alloc_tracker = { path = "../alloc_tracker" }
criterion = { workspace = true }
mutants = { workspace = true }
static_assertions = { workspace = true }

[lints]
workspace = true

[[bench]]
name = "blind_basic"
harness = false

[[bench]]
name = "blind_many_types"
harness = false

[[bench]]
name = "insert_with"
harness = false
