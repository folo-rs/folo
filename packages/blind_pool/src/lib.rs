//! This package provides [`BlindPool`], a dynamically growing pool of objects that can store
//! objects of any type.
//!
//! It offers automatic memory management with stable memory addresses
//! and efficient typed insertion with automatic value dropping.
//!
//! [`LocalBlindPool`] is a single-threaded variant with reduced overhead and [`RawBlindPool`]
//! offers manual lifetime management for the most performance-sensitive scenarios.
//!
//! # Features
//!
//! - **Type-agnostic memory management**: The pool accepts objects of any sized type.
//! - **Automatic resource management**: Pooled objects are dropped automatically when no longer
//!   referenced.
//! - **Thread-safe and single-threaded variants**: [`BlindPool`] for multi-threaded use,
//!   [`LocalBlindPool`] for single-threaded performance.
//! - **Always pinned**: Memory address remains valid for the entire lifetime of a pooled object.
//! - **Dynamic growth**: Pool capacity grows automatically as needed, shrinks to fit on-demand.
//! - **Efficient allocation**: Uses high density slabs to minimize allocation overhead.
//! - **Stable Rust**: No unstable Rust features required.
//! - **Optional leak detection**: Pool can be configured to panic on drop if values are
//!   still present (primarily relevant for [`RawBlindPool`]).
//!
//! # Example
//!
//! ```rust
//! use blind_pool::BlindPool;
//!
//! // Create a thread-safe pool.
//! let pool = BlindPool::new();
//!
//! // Insert values and get handles.
//! let u64_handle = pool.insert(42_u64);
//! let string_handle = pool.insert("hello".to_string());
//!
//! // Access values through dereferencing.
//! assert_eq!(*u64_handle, 42);
//! assert_eq!(*string_handle, "hello");
//!
//! // Values are automatically cleaned up when handles are dropped.
//! ```
//!
//! For single-threaded use:
//!
//! ```rust
//! use blind_pool::LocalBlindPool;
//!
//! // Create a single-threaded pool (more efficient).
//! let pool = LocalBlindPool::new();
//!
//! let value_handle = pool.insert(vec![1, 2, 3]);
//! assert_eq!(*value_handle, vec![1, 2, 3]);
//! ```
//!
//! For manual resource management:
//!
//! ```rust
//! use blind_pool::RawBlindPool;
//!
//! // Create a pool with manual resource management.
//! let mut pool = RawBlindPool::new();
//!
//! // Insert values of different types into the same pool.
//! let pooled_u64 = pool.insert(42_u64);
//! let pooled_i32 = pool.insert(-123_i32);
//! let pooled_f32 = pool.insert(3.14_f32);
//!
//! // Read data back from the pooled items.
//! let value_u64 = unsafe {
//!     // SAFETY: The pointer is valid and the value was just inserted.
//!     *pooled_u64.ptr().as_ref()
//! };
//!
//! let value_i32 = unsafe {
//!     // SAFETY: The pointer is valid and the value was just inserted.
//!     *pooled_i32.ptr().as_ref()
//! };
//!
//! assert_eq!(value_u64, 42);
//! assert_eq!(value_i32, -123);
//!
//! // Manual cleanup required.
//! pool.remove(&pooled_u64);
//! pool.remove(&pooled_i32);
//! pool.remove(&pooled_f32);
//! ```

mod cast;
mod constants;
mod local_pool;
mod local_pooled;
mod pool;
mod pooled;
mod raw;
mod raw_builder;

pub use local_pool::*;
pub use local_pooled::*;
// Re-export DropPolicy from opaque_pool simply because we do not need a different one.
pub use opaque_pool::DropPolicy;
// Re-export so we can use it without the consumer needing a reference.
#[doc(hidden)]
pub use paste::paste as __private_paste;
pub use pool::*;
pub use pooled::*;
pub use raw::*;
pub use raw_builder::*;
