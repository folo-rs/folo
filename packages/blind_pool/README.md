A pinned object pool that can store objects of any type.

```rust
use blind_pool::BlindPool;

// Create a blind pool that can store any type.
let mut pool = BlindPool::new();

// Insert values of different types into the same pool.
let pooled_u64 = pool.insert(42_u64);
let pooled_i32 = pool.insert(-123_i32);
let pooled_f32 = pool.insert(3.14_f32);

// Read data back from the pooled items.
let value_u64 = unsafe {
    // SAFETY: The pointer is valid and the value was just inserted.
    pooled_u64.ptr().read()
};

let value_i32 = unsafe {
    // SAFETY: The pointer is valid and the value was just inserted.
    pooled_i32.ptr().read()
};

assert_eq!(value_u64, 42);
assert_eq!(value_i32, -123);
```

More details in the [package documentation](https://docs.rs/blind_pool/).

This is part of the [Folo project](https://github.com/folo-rs/folo) that provides mechanisms for
high-performance hardware-aware programming in Rust.