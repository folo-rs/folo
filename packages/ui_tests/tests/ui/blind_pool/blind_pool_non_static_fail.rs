// This test should fail to compile because we're trying to insert a non-static reference into BlindPool
use blind_pool::BlindPool;

fn main() {
    let pool = BlindPool::new();
    let value = 42;
    let reference = &value; // This is not 'static
    
    // This should fail to compile due to the 'static requirement
    let _handle = pool.insert(reference);
}