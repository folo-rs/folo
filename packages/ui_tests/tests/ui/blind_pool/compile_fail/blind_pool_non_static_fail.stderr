error[E0597]: `value` does not live long enough
  --> tests/ui/blind_pool/compile_fail/blind_pool_non_static_fail.rs:7:21
   |
6  |     let value = 42;
   |         ----- binding `value` declared here
7  |     let reference = &value; // This is not 'static
   |                     ^^^^^^ borrowed value does not live long enough
...
10 |     let _handle = pool.insert(reference);
   |                   ---------------------- argument requires that `value` is borrowed for `'static`
11 | }
   | - `value` dropped here while still borrowed
